### ISSUE TITLE
Big linking error on windows

### ISSUE THREAD

**Issue by 8176135** [twistedfall/opencv-rust#70]
I am getting a massive linking error ending with 
```
fatal error LNK1120: 106 unresolved externals
```

```
$ echo $OPENCV_LINK_LIBS
opencv_world348     # Also tried with opencv_world348d

$ echo $OPENCV_INCLUDE_PATHS
D:\FDrive\Documents\CStuff\Libraries\opencv3\build\include

$ echo $OPENCV_LINK_PATHS
D:\FDrive\Documents\CStuff\Libraries\opencv3\build\x64\vc14\lib    # Also tried with vc15
```

I'm getting the same error with both powershell and MSYS2

Here is the full error message: https://pastebin.com/LxNk14DK

I have also tried with opencv 4.1.x, and get similar errors.

Any ideas?

 (Btw on my arch linux machine I am also getting similar linking errors, still investigating that)

**Comment by twistedfall** [comment]
Hi, it's all related to viz module and dnn linking issue. Did you build with `buildtime-bindgen` feature?

**Comment by 8176135** [comment]
I just did, and got the same error.

**Comment by twistedfall** [comment]
Can you please share the full log of `cargo run -vv`. You should also try compiling with `buildtime-bindgen` and in release mode

**Comment by 8176135** [comment]
With release, I only get 1 linking error:
```
error: linking with `link.exe` failed: exit code: 1120

...

 note: libopencv-53e045c6d8907fe4.rlib(cpp_closures.o) : error LNK2019: unresolved external symbol "public: virtual __cdecl cv::xfeatures2d::Elliptic_KeyPoint::~Elliptic_KeyPoint(void)" (??1Elliptic_KeyPoint@xfeatures2d@cv@@UEAA@XZ) referenced in function "public: class std::_Vector_iterator<class std::_Vector_val<struct std::_Simple_types<class cv::xfeatures2d::Elliptic_KeyPoint> > > __cdecl std::vector<class cv::xfeatures2d::Elliptic_KeyPoint,class std::allocator<class cv::xfeatures2d::Elliptic_KeyPoint> >::emplace<class cv::xfeatures2d::Elliptic_KeyPoint const &>(class std::_Vector_const_iterator<class std::_Vector_val<struct std::_Simple_types<class cv::xfeatures2d::Elliptic_KeyPoint> > >,class cv::xfeatures2d::Elliptic_KeyPoint const &)" (??$emplace@AEBVElliptic_KeyPoint@xfeatures2d@cv@@@?$vector@VElliptic_KeyPoint@xfeatures2d@cv@@V?$allocator@VElliptic_KeyPoint@xfeatures2d@cv@@@std@@@std@@QEAA?AV?$_Vector_iterator@V?$_Vector_val@U?$_Simple_types@VElliptic_KeyPoint@xfeatures2d@cv@@@std@@@std@@@1@V?$_Vector_const_iterator@V?$_Vector_val@U?$_Simple_types@VElliptic_KeyPoint@xfeatures2d@cv@@@std@@@std@@@1@AEBVElliptic_KeyPoint@xfeatures2d@cv@@@Z)
          C:\Users\Username\Documents\RustProjects\rust_prac\target\release\deps\rust_prac-2519d7032318d987.exe : fatal error LNK1120: 1 unresolved externals
```
Full error: https://pastebin.com/Tw8012mC

and with `cargo run -vv`

https://pastebin.com/La3eNkEH

Running with  `cargo run -vv --release` gives a similar error, but with just 1 unresolved externals like above

**Comment by twistedfall** [comment]
For `cargo run -vv`, can you please do `cargo clean` first? I'd like to see the build log from the opencv crate. BTW, do you have `contrib` feature enabled? If so, can you try building without it?

**Comment by 8176135** [comment]
Well here it is: https://pastebin.com/V6ua8kb7, opencv part starts on line 2926

Here is the cargo.toml in case it matters:

```
[package]
name = "rust_prac"
version = "0.1.0"
authors = ["8176135 <8176135@gmail.com>"]
edition = "2018"

# See more keys and their definitions at https://doc.rust-lang.org/cargo/reference/manifest.html

[dependencies]
opencv = {version = "0.26", default-features = false, features = ["opencv-34", "buildtime-bindgen"]}
```

I haven't tried with contrib yet

**Comment by 8176135** [comment]
Just tried with contrib, the error count goes up to 419

```
fatal error LNK1120: 419 unresolved externals
```

**Comment by twistedfall** [comment]
I can see you have a lot of logging encoding problems too, I'll look into it also. It's not build-critical though. For now please set `OPENCV_HEADER_DIR` environment variable to "D:/FDrive/Documents/CStuff/Libraries/opencv3/build/include", it should help you with the linking problems

**Comment by 8176135** [comment]
Hmm... it reduced the linking errors to just 1 (same error with and without contrib)

```
  = note: libopencv-58f02720fa4c804a.rlib(dnn.o) : error LNK2019: unresolved external symbol "public: __cdecl cv::dnn::experimental_dnn_34_v14::BackendNode::BackendNode(int)" (??0BackendNode@experimental_dnn_34_v14@dnn@cv@@QEAA@H@Z) referenced in function cv_dnn_BackendNode_BackendNode_int
          C:\Users\HoCthulhu\Documents\RustProjects\rust_prac\target\debug\deps\rust_prac-1c909b857dae8a2e.exe : fatal error LNK1120: 1 unresolved externals
```

PS: Just tried with `--release`, and it built, but running crashes with 

```
error: process didn't exit successfully: `target\release\rust_prac.exe` (exit code: 0xc0000135, STATUS_DLL_NOT_FOUND)
```

do I need to manually copy in a dll?

**Comment by twistedfall** [comment]
Well, the system needs to be able to find `opencv_world348.dll` somehow, one way would be to copy it next to the .exe file as far as I know

**Comment by 8176135** [comment]
Ok, that fixed it, the only error now is the linking issue running in debug mode

**Comment by twistedfall** [comment]
Great to hear. The debug mode linking issue in Windows is known, I'm not sure how to fix it at the moment. Just to verify, building with the same settings like with your successful release build, but without `--release` flag still gives you linking errors?

**Comment by 8176135** [comment]
Yes, the only difference between the 2 builds is the release flag

**Comment by twistedfall** [comment]
Generally, travis build for debug version completes successfully so you might want to compare what's different between your config and https://travis-ci.org/twistedfall/opencv-rust/jobs/611780019 If you find the cause please do report it!

**Comment by twistedfall** [comment]
That linking issue should be fixed in 0.28.0 for debug mode too
